### PR TITLE
Provider db

### PR DESCRIPTION
**Pull request by zakpeters**
- add sqlite db
- remove all snapshot dependencies
- error messages

(sorry about the commits from the future, changed my computers clock so that I could check on the activity cycling)

### CODE REVIEW

File: provider/db.go [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Failed to prepare blocks table. error: %s", err)
+	}
+	_, err = stmt.Exec()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create blocks table. error: %s", err)
+	}
+
+	// Create activity table
+	stmt, err = db.Prepare(`CREATE TABLE IF NOT EXISTS activity ( id INTEGER PRIMARY KEY,
```
Comment (AlexSteele) [inline provider/db.go]: The activity handling is clever but, I think, a little wonky. You also lose all of the counts for the previous period when you roll over (correct me if I'm wrong?).

Here's what I recommend you do. I think this is more similar to what you were previously doing:

Keep a top-level struct for the activity feed. In this, keep counters for everything that happened in the last 24 hours on a per-hour basis and in the last 7 days on a per-day basis.

```
type ActivityFeed struct {

	// Counters for the last 24 hours. Typically, len(hourlyCounts) == 7.
	hourlyCounts []activityCounts

	// Daily counts for the last week. Typically, len(dailyCounts) == 7.
	dailyCounts  []activityCounts
}

type activityCounts struct {

	// The timestmap for this set of counters, truncated down to the
	// nearest hour (for hourlyCounters) or day (for daily counters)
	TimeStamp           time.Time
	BlockUploads        int64
	BlockDownloads      int64
	BlockDeletions      int64
	StorageReservations int64
	BytesUploaded       int64
	BytesDownloaded     int64
}
```


Add a method on the feed which updates the counters when a new entry is ready. Take the current timestamp as a parameter to make it easier to test. The method will look roughly like this (there are some edge cases I don't handle correctly, but this is the general gist). 

```
func (feed *ActivityFeed) AddOp(t time.Time, op string, byteAmt int64) {

	// Lop off the old hourly counters from more than 24 hours ago
	idx := 0
	hour := t.Truncate(time.Hour)
	for idx < len(feed.hourlyCounts) && hour.Sub(feed.hourlyCounts[idx].TimeStamp) > 24 * time.Hour {
		idx++
	}
	feed.hourlyCounts = feed.hourlyCounts[idx:]

	// Lop off old daily counters from more than 7 days ago
	idx = 0
	day := t.Truncate(24 * time.Hour)
	for idx < len(feed.dailyCounts) && day.Sub(feed.dailyCounts[idx].TimeStamp) > 7 * 24 * time.Hour {
		idx++
	}
	feed.dailyCounts = feed.dailyCounts[idx:]

	// Append new hourly counters for all the hours between the last counter and now.
	// You'll need to watch for the edge case where len(feed.hourlyCounts) == 0!!!
	for len(feed.hourlyCounts) < 24 {
		lastTimestamp := feed.hourlyCounts[len(feed.hourlyCounts) - 1].TimeStamp
		nextTimestamp := lastTimestamp.Add(time.Hour)
		feed.hourlyCounts = append(feed.hourlyCounts, activityCounts{TimeStamp: nextTimestamp})
	}

	// Ditto here for the daily counters
	for len(feed.dailyCounts) < 7 {

	}

        // Now actually update the counts for the most recent period.
	hourlyCounts := &feed.hourlyCounts[len(feed.hourlyCounts) - 1]
	dailyCounts := &feed.dailyCounts[len(feed.dailyCounts) - 1]
	countsToIncrement := []*activityCounts{hourlyCounts, dailyCounts}

	switch op {
	case "BlockUpload":
		for _, counts := range countsToIncrement {
			counts.BlockUploads++
			counts.BytesUploaded += byteAmt
		}
	//	... And so on and so forth for the different types ...
	}
}
```

Then, add a method to generate the summary expected by the GET /stats endpoint. I'll leave the details to you.

```
func (feed *ActivityFeed) GenerateSummary() *ActivitySummary
```

Then add unit tests for the activity feed.

Finally, DON'T save the activity feed in SQLite. Just put it in a plain old JSON file. Update the code in the cmd/provider.go to handle OS signals. When you get a signal to shutdown, stop the server and save the snapshot file. I've done this a few times before and it looks roughly like this:

```
//  ... inside the provider daemon command
go func() {
  signal := <- os.Signal
  provider.SaveSnapshot()   

  // You'll have to look into this part. Worst case, just call provider.SaveSnapshot() then os.Exit() and don't worry
  // about cleanly shutting down the server.
  server.Shutdown()
  os.Exit(0)
}()
```

File: provider/negotiate.go [hunk 2]
```diff
@@ -27,36 +28,35 @@ func (provider *Provider) NegotiateContract(contract *core.Contract) (*core.Cont
...
+	err = provider.addActivity("contract", contract.StorageSpace)
 	if err != nil {
+		fmt.Println("Failed to update activity for contract: ", err)
+	}
 
-		// TODO: Remove contract. I don't do this here
-		// since we need to move to an improved storage scheme anyways.
-		return nil, fmt.Errorf("Unable to save contract. Error: %s", err)
+	// this could potentially be non-fatal too
+	err = provider.UpdateMeta()
```
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/negotiate.go]: This is just every successful contract creation. I'm not sure where else we would do this as otherwise it's only updated on config update (which really should not happen often).

[the remark this answers is not in the document]
Comment (AlexSteele) [reply provider/negotiate.go]: Here's an idea:

```
func (p *Provider) metaUpdateThread() {
   updateInterval := 5 * time.Minute
   ticker := time.NewTicker(updateInterval)
   for {
      switch {
         case <- tickerC:
             p.updateMeta()
         case <- p.doneCh:
             ticker.Stop()
             return
    }
}

func (p *Provider) StartBackgroundThreads() {
   go p.metaUpdateThread()
}

func (p *Provider) StopBackgroundThreads() {
   close(p.doneCh)
}
```

[the remark this answers is not in the document]
Comment (AlexSteele) [reply provider/negotiate.go]: Or a little more sophisticated: 

```
func (p *Provider) metaUpdateThread() {
   updateInterval := 5 * time.Minute
   ticker := time.NewTicker(updateInterval)
   pinfo := p.GetInfo()
   for {
      switch {
         case <- ticker.C:
             updatedPinfo := p.GetInfo()
             if *updatedPinfo != *pinfo {
                p.updateMeta()
             }
             pinfo = updatedPinfo
         case <- p.doneCh:
             ticker.Stop()
             return
        }
    }
}

func (p *Provider) StartBackgroundThreads() {
   go p.metaUpdateThread()
}

func (p *Provider) StopBackgroundThreads() {
   close(p.doneCh)
}
```

File: provider/db.go (hunk 1 above)
Comment (Caid11) [reply provider/db.go]: Alternatively, I think leveraging SQLite could make things much easier and more robust. Rather than maintaining counters and worrying about locks and such, I would just log all actions to the database, then use database operations (which should be pretty fast) to get the stats we want.

For example, you could use a query like the one mentioned [here](https://stackoverflow.com/questions/10643379/how-do-i-query-for-all-dates-greater-than-a-certain-date-in-sql-server) to get all of the entries newer than a certain day (say, last week), then just use COUNT to get the specific number of that type of action.

I prefer that approach because we don't have to do a lot of complicated logic, let the database do the heavy lifting, and collect more data, which gives us the option to pull more stats out of existing data later if we choose.

File: provider/db.go [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+//  (potentially useful for maintenance also)
+// - provider.StorageReserved
+// - provider.StorageUsed
+// - provider.TotalBlocks
+// - provider.TotalContracts
+// - provider.renters {
+// 	   - StorageUsed
+//     - StorageReserved
+//   }
+func (p *Provider) LoadDBIntoMemory() error {
```
[the remark this answers is not in the document]
Comment (Caid11) [reply provider/db.go]: Agree.

File: provider/db.go [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open DB. error: %s", err)
+	}
+
+	// Create contracts table
+	stmt, err := db.Prepare(`CREATE TABLE IF NOT EXISTS contracts ( id INTEGER PRIMARY KEY, 
+		ContractId TEXT, 
+		RenterId TEXT, 
+		ProviderId TEXT, 
+		StorageSpace INTEGER,
```
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/db.go]: Fixed.

File: provider/db.go (hunk 1 above)
Comment (zakpeters) [reply provider/db.go]: It cycles in the same manner that the old one does, the only difference is that it is stored in SQLite as opposed to a snapshot.

It might be unclear in some of the code (I added some comments) but 'day' intervals correspond to 24 1-hour intervals and they cycle out of scope if they are more than 24-hours old.

I would like to switch to the method Kincaid suggested however I think it would struggle on the UploadedBytes and DownloadedBytes as they would have to be summed as opposed to using the COUNT method.

File: provider/negotiate.go [hunk 5]
```diff
@@ -6,10 +6,11 @@ import (
 	"skybin/core"
 )
 
+// Helper to negotiate contracts
```
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/negotiate.go]: I think it should be concurrency safe just from SQLite (see part 5 of their FAQ). As I understand, it locks the DB for every write op, but reads can occur simultaneously. Please correct me if I'm wrong, I tried to read up on it, but I'm struggling to understand where and when to lock these concurrent operations.

"Multiple processes can have the same database open at the same time. Multiple processes can be doing a SELECT at the same time. But only one process can be making changes to the database at any moment in time, however."

File: provider/server.go [hunk 6]
```diff
@@ -101,7 +103,7 @@ func (server *providerServer) getInfo(w http.ResponseWriter, r *http.Request) {
 		PublicKey:   string(pubKeyBytes),
 		Addr:        server.provider.Config.PublicApiAddr,
 		SpaceAvail:  server.provider.Config.SpaceAvail - server.provider.StorageReserved,
-		StorageRate: 1,
+		StorageRate: server.provider.Config.StorageRate,
```
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/server.go]: The GetInfo method provides the info that the provider dashboard is concerned with such as SpaceAllocated, TotalBlocks, TotalRenters, etc.  

Renter's should only be interested in: ProviderId, Public address, Public Key (Provider dashboard does not need this field), Space Available, and Storage Rate.

File: provider/db.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/db.go]: I think that recalculating these on startup actually makes these fields more resilient (as they can reconstruct from the elements in the DB on application restart if something happens to get out of sync). 

These fields are also incremented in memory in their respective methods.

File: provider/provider.go [hunk 7]
```diff
@@ -27,17 +26,19 @@ type Config struct {
...
-	Homedir    string //move this maybe
+	Homedir string //move this maybe
+	Config  *Config
+	db      *sql.DB
+	mu      sync.Mutex
+
 	PrivateKey *rsa.PrivateKey
-	contracts  []*core.Contract
-	stats      Stats
 	renters    map[string]*RenterInfo
```
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/provider.go]: I did that on my first implementation of the DB, however as per my comment on the LoadDBIntoMemory method. I do think that calculating this map on startup is much more resilient.

File: provider/db.go [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+		contract.ProviderSignature,
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Currently unused, but probably relevant for canceling contracts
+func (p *Provider) DeleteContractkById(renterId string) error {
```
[the remark this answers is not in the document]
Comment (zakpeters) [reply provider/db.go]: Fixed

**Comment by zakpeters** [comment]
* I don't see you setting the id field in the SQLite records anywhere... Maybe SQLite does this automatically?
    * This is auto incremented. From FAQ: “Short answer: A column declared INTEGER PRIMARY KEY will autoincrement.”
* If possible, you may want to add an index to the Blocks and Contracts tables on the fields commonly used for lookup.
    * done
* I might consider breaking the DB logic into a separate object (i.e. type ProviderDB) and adding unit tests for it. I'm nervous merging these without more tests.
    * done
* I also recommend doing the same thing with the Provider object. Implement the logic for the PostBlocks/DeleteBlocks endpoints within methods on this object and just handle parameter validation inside the providerServer, passing the parameters to the appropriate methods for the knitty-gritty work. Then consider adding tests on the provider logic using Go's testing lib.
    * On my todo list
* More on testing: Before we merge this, it would be good to run this for a longer period. Try running either the test_network or new renter_bot python scripts for a few hours with these changes before Tuesday.
    * Ran successfully with test-net for ~10 hours

File: provider/negotiate.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexSteele) [reply provider/negotiate.go]: Right. My concern is around the in-memory fields, like the provider.renters map, the renter info objects, and the provider stats fields. Addressing these should hopefully be pretty simple. First add a sync.Mutex to the provider struct (it may be in there). Then lock around updates to the in-memory fields when they are updated or read in a request handler. Request handlers can run at the same time, including two concurrent executions of the same handler for different requests.

File: provider/db.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexSteele) [reply provider/db.go]: I don't see a real concern with keeping a few integer fields in sync with the DB, as long as they're incremented and decremented correctly. Another potential issue here is that this affects GUI start up time since it must wait for this work to be done before the provider server starts. I approve this for now, though.
